### PR TITLE
WITHOUT ROWID

### PR DESCRIPTION
**Pull request by faithanalog**
Here's the WITHOUT ROWID changes per https://github.com/oxidecomputer/crucible/issues/764 https://github.com/oxidecomputer/crucible/issues/763

Note that as a side effect of this, the contexts/hashes will not be sent to upstairs in the order that they were written, it'll be basically random per the hashing function. This doesn't matter for any blocks read after a flush since there will only be one context anyway. This shouldn't matter for any blocks read before a flush because upstairs already scans thru all contexts and picks the whichever one matches.

So yeah in theory this adds some computational overhead to upstairs for finding the right context since its a random distribution as to how far thru the contexts it needs to scan before finding the right one instead of usually being the first one in the list. But (especially given we're tightening down the flush interval significantly anyway), I think you won't see those context lists get big enough for it to matter.

Also we can revert the integer hashes change if y'all hate that one. it helps with perf a little bit and I like what it does to the code but its on the margins

### CODE REVIEW

**Review by jmpesp — APPROVED** [review]
This is awesome, and produces a noticeable bump during image import. Before:

```
jwm@duster ~/src/oxidecomputer/oxide.rs [main] [] $ ./image.sh
[00:00:22] [████████████████████████████████████████████████████████] 2.00 GiB/2.00 GiB (91.71 MiB/s, 0s)
done!
```

after:

```
jwm@duster ~/src/oxidecomputer/oxide.rs [main] [] $ ./image.sh
[00:00:19] [███████████████████████████████████████████████████████] 2.00 GiB/2.00 GiB (102.42 MiB/s, 0s)
done!
```

File: downstairs/src/region.rs [hunk 1]
```diff
@@ -757,20 +755,34 @@ impl Extent {
...
+            // > WITHOUT ROWID tables work best when individual rows are not too
+            // > large. A good rule-of-thumb is that the average size of a
+            // > single row in a WITHOUT ROWID table should be less than about
+            // > 1/20th the size of a database page. That means that rows should
+            // > not contain more than about 50 bytes each for a 1KiB page size
+            // > or about 200 bytes each for 4KiB page size.
+            //
+            // The default SQLite page size is 4KiB, per https://sqlite.org/pgszchng2016.html
+            //
+            // The primary key is also a uniqueness constraint. Because the
```
[the remark this answers is not in the document]
Comment (faithanalog) [reply downstairs/src/region.rs:771]: yes it is true in WITHOUT ROWID mode and in fact it is the only way that WITHOUT ROWID works. When I first implemented this I had to make the same change to the insertions that you ended up making when you added your row constraint, for this exact reason. sqlite is in some sense just a funny-looking journaled BTreeMap in this mode.

File: downstairs/src/region.rs [hunk 2]
```diff
@@ -757,20 +755,34 @@ impl Extent {
...
+            // on_disk_hash is a hash of the data AFTER encryption, we only need
+            // (block, on_disk_hash). A duplicate write with a different
+            // encryption context necessarily results in a different on disk
+            // hash. Likewise for a write with a different nonce, or a different
+            // nonce.
```
[the remark this answers is not in the document]
Comment (faithanalog) [reply downstairs/src/region.rs:776]: oh yeah i was writing without reading there

**Review by leftwo — APPROVED** [review]
Looks good, just a small nit below.

**Pull request merged**